File: getFileType.java
import java.util.HashMap;
import java.util.Map;

public class FileTypeChecker {
    public enum FileCategory {
        IMAGE, VIDEO, AUDIO, DOCUMENT, ARCHIVE, EXECUTABLE, CODE, FONT, DATABASE, MODEL, DISK_IMAGE, CONFIG, FILE
    }

    public static class FileTypeResult {
        String mimeType;
        FileCategory category;

        public FileTypeResult(String mimeType, FileCategory category) {
            this.mimeType = mimeType;
            this.category = category;
        }
    }

    private static final Map<String, String> mimeTypeMap = new HashMap<>() {{
        put("avif", "image/avif");
        put("jpeg", "image/jpeg");
        put("jpg", "image/jpeg");
        put("png", "image/png");
        put("bmp", "image/bmp");
        put("gif", "image/gif");
        put("svg", "image/svg+xml");
        put("webp", "image/webp");
        put("ico", "image/vnd.microsoft.icon");
        put("tif", "image/tiff");
        put("tiff", "image/tiff");
        put("psd", "image/vnd.adobe.photoshop");
        put("ai", "application/postscript");
        put("sketch", "application/sketch");
        put("mp4", "video/mp4");
        put("webm", "video/webm");
        put("ogv", "video/ogg");
        put("avi", "video/x-msvideo");
        put("mpeg", "video/mpeg");
        put("3gp", "video/3gpp");
        put("3g2", "video/3gpp2");
        put("mp3", "audio/mpeg");
        put("wav", "audio/wav");
        put("weba", "audio/webm");
        put("oga", "audio/ogg");
        put("opus", "audio/opus");
        put("midi", "audio/midi");
        put("aac", "audio/aac");
        put("pdf", "application/pdf");
        put("doc", "application/msword");
        put("docx", "application/vnd.openxmlformats-officedocument.wordprocessingml.document");
        put("xls", "application/vnd.ms-excel");
        put("xlsx", "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
        put("ppt", "application/vnd.ms-powerpoint");
        put("pptx", "application/vnd.openxmlformats-officedocument.presentationml.presentation");
        put("txt", "text/plain");
        put("csv", "text/csv");
        put("rtf", "application/rtf");
        put("odt", "application/vnd.oasis.opendocument.text");
        put("zip", "application/zip");
        put("rar", "application/vnd.rar");
        put("7z", "application/x-7z-compressed");
        put("tar", "application/x-tar");
        put("gz", "application/gzip");
        put("bz2", "application/x-bzip2");
        put("exe", "application/x-msdownload");
        put("sh", "application/x-sh");
        put("bat", "application/x-msdos-program");
        put("apk", "application/vnd.android.package-archive");
        put("ipa", "application/octet-stream");
        put("jar", "application/java-archive");
        put("app", "application/x-macos-app");
        put("deb", "application/x-deb");
        put("rpm", "application/x-rpm");
        put("js", "text/javascript");
        put("ts", "text/typescript");
        put("py", "text/x-python");
        put("java", "text/x-java-source");
        put("cpp", "text/x-c");
        put("c", "text/x-c");
        put("kt", "text/x-kotlin");
        put("rs", "text/x-rust");
        put("swift", "text/x-swift");
        put("go", "text/x-go");
        put("php", "text/x-php");
        put("rb", "text/x-ruby");
        put("sqlite", "application/x-sqlite3");
        put("db", "application/x-database");
        put("sql", "application/sql");
        put("stl", "model/stl");
        put("obj", "model/obj");
        put("fbx", "model/fbx");
        put("blend", "model/blender");
        put("iso", "application/x-iso9660-image");
        put("img", "application/octet-stream");
        put("ini", "text/plain");
        put("log", "text/plain");
        put("cfg", "text/plain");
        put("ttf", "font/ttf");
        put("woff", "font/woff");
        put("woff2", "font/woff2");
        put("otf", "font/otf");
    }};

    private static final Map<String, FileCategory> categoryMap = new HashMap<>() {{
        put("image/", FileCategory.IMAGE);
        put("video/", FileCategory.VIDEO);
        put("audio/", FileCategory.AUDIO);
        put("application/pdf", FileCategory.DOCUMENT);
        put("application/msword", FileCategory.DOCUMENT);
        put("application/vnd.openxmlformats-officedocument", FileCategory.DOCUMENT);
        put("application/vnd.ms-powerpoint", FileCategory.DOCUMENT);
        put("application/vnd.oasis.opendocument", FileCategory.DOCUMENT);
        put("application/vnd.ms-excel", FileCategory.DOCUMENT);
        put("text/plain", FileCategory.DOCUMENT);
        put("text/csv", FileCategory.DOCUMENT);
        put("application/rtf", FileCategory.DOCUMENT);
        put("application/zip", FileCategory.ARCHIVE);
        put("application/vnd.rar", FileCategory.ARCHIVE);
        put("application/x-7z-compressed", FileCategory.ARCHIVE);
        put("application/x-tar", FileCategory.ARCHIVE);
        put("application/gzip", FileCategory.ARCHIVE);
        put("application/x-bzip2", FileCategory.ARCHIVE);
        put("application/x-msdownload", FileCategory.EXECUTABLE);
        put("application/vnd.android.package-archive", FileCategory.EXECUTABLE);
        put("application/octet-stream", FileCategory.EXECUTABLE);
        put("application/java-archive", FileCategory.EXECUTABLE);
        put("application/x-deb", FileCategory.EXECUTABLE);
        put("application/x-rpm", FileCategory.EXECUTABLE);
        put("text/javascript", FileCategory.CODE);
        put("text/typescript", FileCategory.CODE);
        put("text/x-python", FileCategory.CODE);
        put("text/x-java-source", FileCategory.CODE);
        put("text/x-c", FileCategory.CODE);
        put("text/x-kotlin", FileCategory.CODE);
        put("text/x-rust", FileCategory.CODE);
        put("text/x-swift", FileCategory.CODE);
        put("text/x-go", FileCategory.CODE);
        put("text/x-php", FileCategory.CODE);
        put("text/x-ruby", FileCategory.CODE);
        put("application/x-sqlite3", FileCategory.DATABASE);
        put("application/x-database", FileCategory.DATABASE);
        put("application/sql", FileCategory.DATABASE);
        put("model/", FileCategory.MODEL);
        put("application/x-iso9660-image", FileCategory.DISK_IMAGE);
        put("text/plain", FileCategory.CONFIG);
        put("font/", FileCategory.FONT);
    }};

    public static FileTypeResult getFileType(String url) {
        String extension = url.replaceAll(".*\\.([0-9a-z]+).*", "$1").toLowerCase();
        String mimeType = mimeTypeMap.getOrDefault(extension, "application/octet-stream");

        FileCategory category = categoryMap.get(mimeType);
        if (category == null) {
            category = categoryMap.entrySet().stream()
                    .filter(entry -> mimeType.startsWith(entry.getKey()))
                    .map(Map.Entry::getValue)
                    .findFirst()
                    .orElse(FileCategory.FILE);
        }

        return new FileTypeResult(mimeType, category);
    }

    public static void main(String[] args) {
        System.out.println(getFileType("example.jpg").mimeType);
    }
}

File: getFileType.py
import re

class FileCategory:
    IMAGE = "image"
    VIDEO = "video"
    AUDIO = "audio"
    DOCUMENT = "document"
    ARCHIVE = "archive"
    EXECUTABLE = "executable"
    CODE = "code"
    FONT = "font"
    DATABASE = "database"
    MODEL = "3d-model"
    DISK_IMAGE = "disk-image"
    CONFIG = "config"
    FILE = "file"

mime_type_map = {
    "avif": "image/avif", "jpeg": "image/jpeg", "jpg": "image/jpeg", "png": "image/png",
    "bmp": "image/bmp", "gif": "image/gif", "svg": "image/svg+xml", "webp": "image/webp",
    "ico": "image/vnd.microsoft.icon", "tif": "image/tiff", "tiff": "image/tiff",
    "psd": "image/vnd.adobe.photoshop", "ai": "application/postscript", "sketch": "application/sketch",
    "mp4": "video/mp4", "webm": "video/webm", "ogv": "video/ogg", "avi": "video/x-msvideo",
    "mpeg": "video/mpeg", "3gp": "video/3gpp", "3g2": "video/3gpp2", "mp3": "audio/mpeg",
    "wav": "audio/wav", "weba": "audio/webm", "oga": "audio/ogg", "opus": "audio/opus",
    "midi": "audio/midi", "aac": "audio/aac", "pdf": "application/pdf", "doc": "application/msword",
    "docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document", "xls": "application/vnd.ms-excel",
    "xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet", "ppt": "application/vnd.ms-powerpoint",
    "pptx": "application/vnd.openxmlformats-officedocument.presentationml.presentation", "txt": "text/plain",
    "csv": "text/csv", "rtf": "application/rtf", "odt": "application/vnd.oasis.opendocument.text",
    "zip": "application/zip", "rar": "application/vnd.rar", "7z": "application/x-7z-compressed",
    "tar": "application/x-tar", "gz": "application/gzip", "bz2": "application/x-bzip2", "exe": "application/x-msdownload",
    "sh": "application/x-sh", "bat": "application/x-msdos-program", "apk": "application/vnd.android.package-archive",
    "ipa": "application/octet-stream", "jar": "application/java-archive", "app": "application/x-macos-app",
    "deb": "application/x-deb", "rpm": "application/x-rpm", "js": "text/javascript", "ts": "text/typescript",
    "py": "text/x-python", "java": "text/x-java-source", "cpp": "text/x-c", "c": "text/x-c", "kt": "text/x-kotlin",
    "rs": "text/x-rust", "swift": "text/x-swift", "go": "text/x-go", "php": "text/x-php", "rb": "text/x-ruby",
    "sqlite": "application/x-sqlite3", "db": "application/x-database", "sql": "application/sql", "stl": "model/stl",
    "obj": "model/obj", "fbx": "model/fbx", "blend": "model/blender", "iso": "application/x-iso9660-image",
    "img": "application/octet-stream", "ini": "text/plain", "log": "text/plain", "cfg": "text/plain",
    "ttf": "font/ttf", "woff": "font/woff", "woff2": "font/woff2", "otf": "font/otf"
}

category_map = {
    "image/": FileCategory.IMAGE, "video/": FileCategory.VIDEO, "audio/": FileCategory.AUDIO,
    "application/pdf": FileCategory.DOCUMENT, "application/msword": FileCategory.DOCUMENT,
    "application/vnd.openxmlformats-officedocument": FileCategory.DOCUMENT, "application/vnd.ms-powerpoint": FileCategory.DOCUMENT,
    "application/vnd.oasis.opendocument": FileCategory.DOCUMENT, "application/vnd.ms-excel": FileCategory.DOCUMENT,
    "text/plain": FileCategory.DOCUMENT, "text/csv": FileCategory.DOCUMENT, "application/rtf": FileCategory.DOCUMENT,
    "application/zip": FileCategory.ARCHIVE, "application/vnd.rar": FileCategory.ARCHIVE, "application/x-7z-compressed": FileCategory.ARCHIVE,
    "application/x-tar": FileCategory.ARCHIVE, "application/gzip": FileCategory.ARCHIVE, "application/x-bzip2": FileCategory.ARCHIVE,
    "application/x-msdownload": FileCategory.EXECUTABLE, "application/vnd.android.package-archive": FileCategory.EXECUTABLE,
    "application/octet-stream": FileCategory.EXECUTABLE, "application/java-archive": FileCategory.EXECUTABLE,
    "application/x-deb": FileCategory.EXECUTABLE, "application/x-rpm": FileCategory.EXECUTABLE,
    "text/javascript": FileCategory.CODE, "text/typescript": FileCategory.CODE, "text/x-python": FileCategory.CODE,
    "text/x-java-source": FileCategory.CODE, "text/x-c": FileCategory.CODE, "text/x-kotlin": FileCategory.CODE,
    "text/x-rust": FileCategory.CODE, "text/x-swift": FileCategory.CODE, "text/x-go": FileCategory.CODE,
    "text/x-php": FileCategory.CODE, "text/x-ruby": FileCategory.CODE, "application/x-sqlite3": FileCategory.DATABASE,
    "application/x-database": FileCategory.DATABASE, "application/sql": FileCategory.DATABASE, "model/": FileCategory.MODEL,
    "application/x-iso9660-image": FileCategory.DISK_IMAGE, "text/plain": FileCategory.CONFIG, "font/": FileCategory.FONT
}

def get_file_type(url: str) -> dict:
    extension = re.search(r"\.([0-9a-z]+)", url)
    if not extension:
        return {"mimeType": "unknown", "category": FileCategory.FILE}

    extension = extension.group(1).lower()
    mime_type = mime_type_map.get(extension, "application/octet-stream")

    category = category_map.get(mime_type)
    if not category:
        category = next((value for key, value in category_map.items() if mime_type.startswith(key)), FileCategory.FILE)

    return {"mimeType": mime_type, "category": category}

# Test
print(get_file_type("example.jpg")["mimeType"])
